[CVP] Implement type narrowing for LShr

adam-bzowski opened the pull request:
Implements type narrowing for LShr. The treatment is analogous to the type narrowing of UDiv. Since LShr is a relatively cheap instruction, the narrowing occurs only if the following conditions hold: i) all the users of the LShr instruction are already TruncInst; ii) the narrowing is carried out to the largest TruncInst following the LShr instruction. Additionally, the function optimizes the cases where the result of the LShr instruction is guaranteed to vanish or be equal to poison.

adam-bzowski committed:
[CVP] Implement type narrowing for LShr

Implements type narrowing for LShr. The treatment is analogous to the type narrowing of UDiv. Since LShr is a relatively cheap instruction, the narrowing occurs only if the following conditions hold: i) all the users of the LShr instruction are already TruncInst; ii) the narrowing is carried out to the largest TruncInst following the LShr instruction. Additionally, the function optimizes the cases where the result of the LShr instruction is guaranteed to vanish or be equal to poison.

dtcxzyw commented on llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:
```diff
@@ -1067,6 +1074,124 @@ static bool processSDivOrSRem(BinaryOperator *Instr, LazyValueInfo *LVI) {
+
+  // What is the smallest bit width that can accommodate the entire value ranges
+  // of both of the operands? Don't shrink below 8 bits wide.
+  unsigned NewWidth = std::max<unsigned>(PowerOf2Ceil(MaxActiveBits), 8);
```
We should avoid narrowing the type to an illegal type. See also `InstCombinerImpl::shouldChangeType`.

adam-bzowski replied:
This part is taken from narrowUDivOrURem and narrowSDivOrSRem. It does look suspicious, though. I can correct in all 3 places or leave it.

dtcxzyw commented on llvm/test/Transforms/CorrelatedValuePropagation/lshr-plus-instcombine.ll:
```diff
@@ -0,0 +1,114 @@
+; CHECK-NEXT: [[C1:%.*]] = lshr i16 [[CARG]], [[CSHIFT]]
+; CHECK-NEXT: ret i16 [[C1]]
+
+define i16 @trunc_test1(i32 %a, i32 %b) {
```
Can you demonstrate that this patch will improve the codegen? https://godbolt.org/z/hbP9qrsxv

BTW this transformation is incorrect: https://alive2.llvm.org/ce/z/j2w5Wa

adam-bzowski replied:
1. The original motivation is a subtle difference between the division and right shift. Due to the lack of narrowing for LShr, UDivs may produce a better code, here's an example:

https://godbolt.org/z/7EPnj9a8d

With the proposed patch the two functions generate identical (better) code.

2. The incorrect transformation was fixed by checking if the maximal shift is smaller than the new bit width.

dtcxzyw commented on llvm/test/Transforms/CorrelatedValuePropagation/lshr-plus-instcombine.ll:
```diff
@@ -0,0 +1,114 @@
+; CHECK-NEXT: [[C1:%.*]] = lshr i16 [[A:%.*]], 2
+; CHECK-NEXT: ret i16 [[C1]]
+
+define i16 @trunc_test2(i16 %a) {
```
It has been handled by InstCombine (`SimplifyDemandedBits`?): https://godbolt.org/z/143fs31s4

dtcxzyw reviewed (changes requested):
It is incorrect. You should check that the shift amount is always smaller than `NewWidth`.

nikic reviewed (commented):
I don't think CVP should be doing this transform. This looks like the job of TruncInstCombine in AggressiveInstCombine?

adam-bzowski committed:
[CVP] Implement type narrowing for LShr

Implements type narrowing for LShr. The treatment is analogous to the type narrowing of UDiv. Since LShr is a relatively cheap instruction, the narrowing occurs only if the following conditions hold: i) all the users of the LShr instruction are already TruncInst; ii) the narrowing is carried out to the largest TruncInst following the LShr instruction. Additionally, the function optimizes the cases where the result of the LShr instruction is guaranteed to vanish or be equal to poison.

adam-bzowski commented:
> It is incorrect. You should check that the shift amount is always smaller than `NewWidth`.

Good point. Fixed that and added some LIT tests to check,

adam-bzowski committed:
[CVP] Implement type narrowing for LShr

Implements type narrowing for LShr. The treatment is analogous to the type narrowing of UDiv. Since LShr is a relatively cheap instruction, the narrowing occurs only if the following conditions hold: i) all the users of the LShr instruction are already TruncInst; ii) the narrowing is carried out to the largest TruncInst following the LShr instruction. Additionally, the function optimizes the cases where the result of the LShr instruction is guaranteed to vanish or be equal to poison.

adam-bzowski committed:
[CVP] Implement type narrowing for LShr

Implements type narrowing for LShr. The treatment is analogous to the type narrowing of UDiv. Since LShr is a relatively cheap instruction, the narrowing occurs only if the following conditions hold: i) all the users of the LShr instruction are already TruncInst; ii) the narrowing is carried out to the largest TruncInst following the LShr instruction. Additionally, the function optimizes the cases where the result of the LShr instruction is guaranteed to vanish or be equal to poison.

adam-bzowski committed:
[CVP] Implement type narrowing for LShr

Implements type narrowing for LShr. The treatment is analogous to the type narrowing of UDiv. Since LShr is a relatively cheap instruction, the narrowing occurs only if the following conditions hold: i) all the users of the LShr instruction are already TruncInst; ii) the narrowing is carried out to the largest TruncInst following the LShr instruction. Additionally, the function optimizes the cases where the result of the LShr instruction is guaranteed to vanish or be equal to poison.

adam-bzowski commented:
> I don't think CVP should be doing this transform. This looks like the job of TruncInstCombine in AggressiveInstCombine?

I don't think it would work. TruncInstCombine doesn't use LazyValueInfo, which is essential here. That's why it fits CVP more, which already deals with similar narrowings.

nikic commented:
It doesn't use LVI, but it does use KnownBits. Why do you need LVI?

nikic commented:
Generally, can you share what your actual motivating case here is?

dtcxzyw commented:
> Generally, can you share what your actual motivating case here is?

CVP narrows UDiv but not LShr.

adam-bzowski commented:
> We already have various optimizations that do this, e.g. see canEvaluateTruncated in InstCombine and TruncInstCombine in AggressiveInstCombine. I think the reason they don't work for your sample is that computeKnownBits probably fails to determine that the top bits are zero in your sample. Probably it's just a matter of adding support for isSignedMinMaxIntrinsicClamp in computeKnownBits. (If we didn't fold the ashr to lshr it would probably also work via ComputeNumSignBits, which does handle the clamp pattern.)

I'll have a look there tomorrow and see if I can implement the narrowing in TruncInstCombine. I agree that it's not specific to lshr. Thanks!

adam-bzowski commented:
I looked at TruncInstCombine as suggested and, indeed, I can make very small changes there (6 lines of code added), which are beneficial even in a wider context than discussed here. I opened a new PR, here:
https://github.com/llvm/llvm-project/pull/120576
I would appreciate any comments and thanks for your help.
